core/src/main/java/jenkins/install/StartupUtil.java

```diff
@@ -0,0 +1,159 @@
+/*
+ * The MIT License
+ *
+ * Copyright (c) 2015, CloudBees, Inc.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package jenkins.install;
+
+import hudson.Functions;
+import hudson.util.VersionNumber;
+import jenkins.model.Jenkins;
+import jenkins.util.xml.XMLUtils;
+import org.apache.commons.io.FileUtils;
+
+import javax.annotation.Nonnull;
+import java.io.File;
+import java.io.IOException;
+import java.util.logging.Logger;
+
+import static java.util.logging.Level.SEVERE;
+
+/**
+ * Jenkins startup utilities.
+ *
+ * @author <a href="mailto:tom.fennelly@gmail.com">tom.fennelly@gmail.com</a>
+ * @since FIXME
+ */
+public class StartupUtil {
+
+    private static final Logger LOGGER = Logger.getLogger(StartupUtil.class.getName());
+
+    private static final VersionNumber NEW_INSTALL_VERSION = new VersionNumber("1.0");
+    
+    /**
+     * Get the type of "startup" currently under way in Jenkins.
+     * @return The type of "startup" currently under way in Jenkins.
+     */
+    public static StartupType getStartupType() {
+        if (Functions.getIsUnitTest()) {
+            return StartupType.TEST;
+        }
+        
+        VersionNumber lastRunVersion = new VersionNumber(getLastExecVersion());
+
+        // Neither the top level config or the lastExecVersionFile have a version
+        // stored in them, which means it's a new install.
+        if (lastRunVersion.compareTo(NEW_INSTALL_VERSION) == 0) {
```

Comment:
Yeah maybe. I was following on from the fact that Jenkins already had the concept of a "1.0" version for a jenkins instance that never had its global config saved.